File: task_reminders.py
"""Private in-app task acknowledgements; live references, never provider writes."""
from datetime import date, datetime, time, timedelta, timezone
import hashlib
import json
import re
from zoneinfo import ZoneInfo

from flask import jsonify, request

from finance_source_bridge import ImportSession
import task_dependencies

TIME_ZONE = 'Asia/Shanghai'
PAGE_SIZE = 40
MAX_REVISION = 9007199254740991
MAX_STATES = 20000
MAX_OPERATIONS = 40000
HEARTBEAT = 'task-reminders-worker'
SCHEMA_SQL = '''
CREATE TABLE IF NOT EXISTS task_reminders(
    owner TEXT NOT NULL REFERENCES users(id), task_id TEXT NOT NULL, due TEXT NOT NULL,
    read_at TEXT, snoozed_until TEXT, revision INTEGER NOT NULL DEFAULT 0 CHECK(revision>=0),
    active INTEGER NOT NULL DEFAULT 1 CHECK(active IN (0,1)),
    created_at TEXT NOT NULL, updated_at TEXT NOT NULL,
    PRIMARY KEY(owner,task_id,due));
CREATE TABLE IF NOT EXISTS task_reminder_operations(
    owner TEXT NOT NULL REFERENCES users(id), request_id TEXT NOT NULL,
    intent_digest TEXT NOT NULL, result TEXT NOT NULL, created_at TEXT NOT NULL,
    PRIMARY KEY(owner,request_id));
'''


class ReminderError(ValueError):
    def __init__(self, message, status=400, code='invalid_reminder_request'):
        super().__init__(message)
        self.message, self.status, self.code = message, status, code


def clock():
    return datetime.now(timezone.utc)


def iso(value):
    return value.astimezone(timezone.utc).isoformat(timespec='microseconds')


def packed(value):
    return json.dumps(value, sort_keys=True, ensure_ascii=False, separators=(',', ':'))


def occurrence(task_id, due):
    return hashlib.sha256(packed(['task-reminder-v1', task_id, due]).encode()).hexdigest()


def clear_member(con, owner):
    """Membership stop owns the transaction; never clear on ordinary logout."""
    if not con.in_transaction:
        raise RuntimeError('reminder_cleanup_requires_transaction')
    # Membership domain also supports a pre-reminder household during migration.
    existing = {r[0] for r in con.execute("SELECT name FROM sqlite_master WHERE type='table'")}
    for table in ('task_reminders', 'task_reminder_operations'):
        if table in existing:
            con.execute('DELETE FROM ' + table + ' WHERE owner=?', (owner,))


def live_tasks(con, moment):
    """Current household shared task ACL plus narrower recipient routing.

    Selected task sources are explicitly shared by cloud_accounts.select().
    A cloud mirror also needs its actual source/item/account and active owner;
    stale or forged JSON sync data cannot grant reminder access.
    """
    members = {r[0] for r in con.execute("SELECT member_id FROM household_memberships WHERE state='active'")}
    cloud = {r['entity_id']: dict(r) for r in con.execute(
        "SELECT i.entity_id,i.source_id,i.remote_id,a.provider FROM cloud_items i "
        "JOIN cloud_sources s ON s.id=i.source_id JOIN cloud_accounts a ON a.id=s.account_id "
        "JOIN household_memberships m ON m.member_id=a.owner "
        "WHERE s.kind='tasks' AND s.owner='shared' AND m.state='active'")}
    graph = task_dependencies.task_graph(con)
    result = {}
    for row in con.execute("SELECT id,data,revision FROM entities WHERE kind='tasks'"):
        value = json.loads(row['data'])
        due = value.get('due')
        if value.get('done') is True or not isinstance(due, str) or not re.fullmatch(r'\d{4}-\d{2}-\d{2}', due):
            continue
        try:
            day = date.fromisoformat(due)
        except ValueError:
            continue
        if not 2000 <= day.year <= 2100:
            continue
        eligible = datetime.combine(day, time(9), ZoneInfo(TIME_ZONE))
        if moment < eligible:
            continue
        owner = value.get('owner')
        recipients = members if owner == 'shared' else ({owner} & members if isinstance(owner, str) else set())
        sync = value.get('sync')
        if sync:
            known = cloud.get(row['id'])
            if (not isinstance(sync, dict) or not known or owner != 'shared'
                    or any(sync.get(k) != known[v] for k, v in
                           (('sourceId', 'source_id'), ('remoteId', 'remote_id'), ('provider', 'provider')))):
                continue
        if not recipients:
            continue
        projected = task_dependencies.project(value, graph)
        task = {k: projected.get(k) for k in ('title', 'owner', 'due', 'dependsOn', 'blockedBy', 'dependencyStatus')}
        task.update(id=row['id'], revision=row['revision'])
        if sync:
            task['sync'] = {k: sync[k] for k in ('provider', 'sourceId', 'remoteId', 'version', 'readOnly') if k in sync}
        result[row['id']] = {'task': task, 'recipients': recipients, 'eligibleAt': iso(eligible),
                             'occurrence': occurrence(row['id'], due)}
    return result


def state_value(state, moment):
    read_at = state['read_at'] if state else None
    snooze = state['snoozed_until'] if state else None
    status = 'snoozed' if snooze and datetime.fromisoformat(snooze) > moment else 'read' if read_at else 'unread'
    return {'revision': state['revision'] if state else 0, 'readAt': read_at,
            'snoozedUntil': snooze, 'status': status}


def capacity(con, table, maximum):
    if con.execute('SELECT count(*) FROM ' + table).fetchone()[0] >= maximum:
        raise ReminderError('提醒历史已达上限，未删除已有记录，请稍后处理', 409, 'reminder_capacity')


class ReminderEngine:
    def __init__(self, app, db):
        self.app, self.db = app, db

    def tick(self):
        """Local transactional materialization, independent of HTTP identities."""
        with self.app.app_context():
            con = self.db()
            try:
                con.execute('BEGIN IMMEDIATE')
                moment = clock(); stamp = iso(moment)
                tasks = live_tasks(con, moment)
                # Inactive history is retained for same-date reopening. Removed
                # memberships are the only automatic destructive history cleanup.
                for table in ('task_reminders', 'task_reminder_operations'):
                    con.execute('DELETE FROM ' + table + " WHERE owner NOT IN (SELECT member_id FROM household_memberships WHERE state='active')")
                stored = {(r['owner'], r['task_id'], r['due']): r for r in con.execute('SELECT * FROM task_reminders')}
                for key, row in stored.items():
                    owner, uid, due = key
                    task = tasks.get(uid)
                    active = int(bool(task and owner in task['recipients'] and task['task']['due'] == due))
                    if row['active'] != active:
                        con.execute('UPDATE task_reminders SET active=?,updated_at=? WHERE owner=? AND task_id=? AND due=?',
                                    (active, stamp, *key))
                room = max(0, MAX_STATES - len(stored)); new = []; blocked = False
                # Bound new combinations by available capacity; never allocate
                # the full task x member cross product just to discard it.
                for uid, task in sorted(tasks.items()):
                    for owner in sorted(task['recipients']):
                        key = (owner, uid, task['task']['due'])
                        if key not in stored:
                            if len(new) >= room:
                                blocked = True
                                break
                            new.append(key)
                    if blocked: break
                for owner, uid, due in new:
                    con.execute('INSERT INTO task_reminders(owner,task_id,due,created_at,updated_at) VALUES(?,?,?,?,?)',
                                (owner, uid, due, stamp, stamp))
                heartbeat = {'lastCheckedAt': stamp, 'capacityBlocked': blocked}
                current = con.execute('SELECT data FROM settings WHERE id=?', (HEARTBEAT,)).fetchone()
                if not current or json.loads(current['data']) != heartbeat:
                    con.execute('INSERT INTO settings(id,data) VALUES(?,?) ON CONFLICT(id) DO UPDATE SET data=excluded.data,revision=revision+1',
                                (HEARTBEAT, packed(heartbeat)))
                con.commit()
                return {'examined': len(tasks), 'generated': min(room, len(new)), **heartbeat}
            except BaseException:
                con.rollback()
                raise


def register_task_reminders(app, db, Problem, body, require_member):
    with app.app_context():
        con = db()
        try:
            con.executescript('BEGIN IMMEDIATE;\n' + SCHEMA_SQL + '\nCOMMIT;')
        except BaseException:
            con.rollback()
            raise
    app.extensions['task_reminders'] = ReminderEngine(app, db)

    @app.errorhandler(ReminderError)
    def reminder_error(error):
        return jsonify(error=error.message, code=error.code), error.status

    def session():
        return ImportSession(app, db, Problem, require_member)

    def hex_id(value, size):
        if not isinstance(value, str) or not re.fullmatch('[0-9a-f]{' + str(size) + '}', value):
            raise ReminderError('提醒或操作编号格式不正确')
        return value

    @app.get('/api/task-reminders')
    def inbox():
        if set(request.args) - {'filter', 'page'} or any(len(request.args.getlist(k)) != 1 for k in request.args):
            raise ReminderError('提醒筛选参数不正确')
        mode, page = request.args.get('filter', 'unread'), request.args.get('page', '0')
        if mode not in ('unread', 'all') or not re.fullmatch(r'0|[1-9][0-9]{0,5}', page):
            raise ReminderError('提醒筛选参数不正确')
        page = int(page); access = session()
        with access.read() as con:
            moment = clock()
            states = {(r['task_id'], r['due']): r for r in con.execute('SELECT * FROM task_reminders WHERE owner=?', (access.owner,))}
            items = []
            for uid, value in live_tasks(con, moment).items():
                if access.owner not in value['recipients']:
                    continue
                items.append({k: value[k] for k in ('task', 'occurrence', 'eligibleAt')} |
                             state_value(states.get((uid, value['task']['due'])), moment))
            items.sort(key=lambda item: (item['task']['due'], item['task']['id']))
            unread = sum(item['status'] == 'unread' for item in items)
            if mode == 'unread': items = [item for item in items if item['status'] == 'unread']
            row = con.execute('SELECT data FROM settings WHERE id=?', (HEARTBEAT,)).fetchone()
            worker = json.loads(row['data']) if row else {'lastCheckedAt': None, 'capacityBlocked': False}
            last = worker['lastCheckedAt']
            worker['stale'] = last is None or not 0 <= (moment - datetime.fromisoformat(last)).total_seconds() <= 60
            result = {'items': items[page*PAGE_SIZE:(page+1)*PAGE_SIZE], 'page': page, 'pageSize': PAGE_SIZE,
                      'total': len(items), 'hasMore': (page+1)*PAGE_SIZE < len(items), 'unreadCount': unread,
                      'serverNow': iso(moment), 'timeZone': TIME_ZONE, 'worker': worker}
        return jsonify(result)

    def normalize(task_id, value):
        if not task_id or len(task_id) > 128 or any(ord(c) < 32 for c in task_id):
            raise ReminderError('任务编号不正确')
        required = {'requestId', 'occurrence', 'revision', 'action'}
        if not isinstance(value, dict) or set(value) not in (required, required | {'snoozedUntil'}):
            raise ReminderError('提醒操作字段不正确')
        hex_id(value['requestId'], 32); hex_id(value['occurrence'], 64)
        if type(value['revision']) is not int or not 0 <= value['revision'] < MAX_REVISION:
            raise ReminderError('提醒版本不正确')
        if value['action'] not in ('read', 'snooze'):
            raise ReminderError('提醒操作不正确')
        result = {**value, 'taskId': task_id}
        if value['action'] == 'read':
            if 'snoozedUntil' in value: raise ReminderError('已读操作不接受暂缓时间')
        else:
            try:
                until = value['snoozedUntil']
                if not isinstance(until, str) or len(until) > 40: raise ValueError()
                parsed = datetime.fromisoformat(until)
                if parsed.tzinfo is None: raise ValueError()
                result['snoozedUntil'] = iso(parsed)
            except (ValueError, KeyError, OverflowError):
                raise ReminderError('暂缓时间须为带时区的有效时间') from None
        return result

    @app.post('/api/task-reminders/<task_id>/actions')
    def action(task_id):
        access = session(); value = normalize(task_id, body())
        intent = hashlib.sha256(packed(value).encode()).hexdigest()
        with access.write() as con:
            old = con.execute('SELECT intent_digest,result FROM task_reminder_operations WHERE owner=? AND request_id=?',
                              (access.owner, value['requestId'])).fetchone()
            if old:
                if old['intent_digest'] != intent:
                    raise ReminderError('此操作编号已用于其他内容', 409, 'reminder_request_conflict')
                result = json.loads(old['result'])
            else:
                moment = clock(); stamp = iso(moment)
                live = live_tasks(con, moment).get(task_id)
                if not live or access.owner not in live['recipients']:
                    raise ReminderError('提醒已不适用，请重新读取', 409, 'reminder_unavailable')
                if live['occurrence'] != value['occurrence']:
                    raise ReminderError('任务截止日期已变化，请重新读取', 409, 'reminder_stale')
                due = live['task']['due']
                key = (access.owner, task_id, due)
                state = con.execute('SELECT * FROM task_reminders WHERE owner=? AND task_id=? AND due=?', key).fetchone()
                if (state['revision'] if state else 0) != value['revision']:
                    raise ReminderError('提醒已更新，请重新读取', 409, 'reminder_stale')
                snooze = value.get('snoozedUntil')
                if snooze and not moment < datetime.fromisoformat(snooze) <= moment + timedelta(days=30):
                    raise ReminderError('暂缓时间须在未来 30 天内')
                capacity(con, 'task_reminder_operations', MAX_OPERATIONS)
                if not state:
                    capacity(con, 'task_reminders', MAX_STATES)
                    con.execute('INSERT INTO task_reminders(owner,task_id,due,created_at,updated_at) VALUES(?,?,?,?,?)', (*key, stamp, stamp))
                read_at = stamp if value['action'] == 'read' else None
                con.execute('UPDATE task_reminders SET read_at=?,snoozed_until=?,revision=revision+1,active=1,updated_at=? '
                            'WHERE owner=? AND task_id=? AND due=?', (read_at, snooze, stamp, *key))
                result = {'requestId': value['requestId'], 'action': value['action'], 'taskId': task_id,
                          'occurrence': value['occurrence'], 'revision': value['revision']+1,
                          'readAt': read_at, 'snoozedUntil': snooze, 'committedAt': stamp}
                con.execute('INSERT INTO task_reminder_operations(owner,request_id,intent_digest,result,created_at) VALUES(?,?,?,?,?)',
                            (access.owner, value['requestId'], intent, packed(result), stamp))
        return jsonify(operation=result)

    @app.get('/api/task-reminders/operations/<request_id>')
    def operation(request_id):
        hex_id(request_id, 32); access = session()
        with access.read() as con:
            row = con.execute('SELECT result FROM task_reminder_operations WHERE owner=? AND request_id=?',
                              (access.owner, request_id)).fetchone()
            if not row:
                raise ReminderError('没有此操作回执', 404, 'reminder_operation_not_found')
            result = json.loads(row['result'])
        return jsonify(operation=result)

File: frontend/src/lib/taskReminders.ts
import type {ListItem} from './types';

export type ReminderFilter='unread'|'all';
export type ReminderTask=Pick<ListItem,'id'|'title'|'owner'|'due'|'revision'|'dependsOn'|'blockedBy'|'dependencyStatus'|'sync'>;
export type TaskReminder={task:ReminderTask;occurrence:string;revision:number;eligibleAt:string;status:'unread'|'read'|'snoozed';readAt:string|null;snoozedUntil:string|null};
export type ReminderPage={items:TaskReminder[];unreadCount:number;total:number;page:number;pageSize:40;hasMore:boolean;serverNow:string;timeZone:'Asia/Shanghai';worker:{lastCheckedAt:string|null;stale:boolean;capacityBlocked?:boolean}};
export type ReminderIntent={taskId:string;requestId:string;occurrence:string;revision:number;action:'read'|'snooze';snoozedUntil?:string};
export type ReminderOperation=Omit<ReminderIntent,'snoozedUntil'>&{readAt:string|null;snoozedUntil:string|null;committedAt:string};
export type ReminderRecovery={intent:ReminderIntent;filter:ReminderFilter;page:number};
const bad=():never=>{throw new Error('提醒资料暂时无法核对，请重新读取。');};
const obj=(v:unknown):Record<string,unknown>=>v!==null&&typeof v==='object'&&!Array.isArray(v)?v as Record<string,unknown>:bad();
const text=(v:unknown,max=128):string=>typeof v==='string'&&v.length>0&&v.length<=max?v:bad();
const taskTitle=(v:unknown):string=>typeof v==='string'&&v.length>0&&Array.from(v).length<=2048?v:bad();
const integer=(v:unknown,min=0):number=>typeof v==='number'&&Number.isSafeInteger(v)&&v>=min?v:bad();
const hex=(v:unknown,len:number)=>typeof v==='string'&&new RegExp('^[a-f0-9]{'+len+'}$').test(v)?v:bad();
const stamp=(v:unknown):string=>typeof v==='string'&&/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,6})?(Z|[+-]\d{2}:\d{2})$/.test(v)&&Number.isFinite(Date.parse(v))?v:bad();
const nullableStamp=(v:unknown)=>v===null?null:stamp(v);
const flag=(v:unknown)=>typeof v==='boolean'?v:bad();
const ids=(v:unknown)=>{if(!Array.isArray(v)||v.length>20)bad();const result=(v as unknown[]).map(x=>text(x));if(new Set(result).size!==result.length)bad();return result;};
export const reminderQuery=(filter:ReminderFilter,page:number)=>'/task-reminders?filter='+filter+'&page='+integer(page);
export function readReminderPage(raw:unknown,filter:ReminderFilter,page:number,memberId:string):ReminderPage {
  const v=obj(raw),now=stamp(v.serverNow),worker=obj(v.worker);
  if(v.page!==page||v.pageSize!==40||v.timeZone!=='Asia/Shanghai'||!Array.isArray(v.items)||v.items.length>40)bad();
  const items=(v.items as unknown[]).map(raw=>{
    const row=obj(raw),t=obj(row.task),due=text(t.due,10);
    if(!/^\d{4}-\d{2}-\d{2}$/.test(due)||!Number.isFinite(Date.parse(due+'T00:00:00Z'))||new Date(due+'T00:00:00Z').toISOString().slice(0,10)!==due||![memberId,'shared'].includes(String(t.owner)))bad();
    if(!['ready','blocked','done'].includes(String(t.dependencyStatus))||!['unread','read','snoozed'].includes(String(row.status))||filter==='unread'&&row.status!=='unread')bad();
    const task:ReminderTask={id:text(t.id),title:taskTitle(t.title),owner:text(t.owner),due,revision:integer(t.revision,1),dependsOn:ids(t.dependsOn),blockedBy:ids(t.blockedBy),dependencyStatus:t.dependencyStatus as ReminderTask['dependencyStatus']};
    if(t.sync!==undefined)task.sync=obj(t.sync);
    const result:TaskReminder={task,occurrence:hex(row.occurrence,64),revision:integer(row.revision),eligibleAt:stamp(row.eligibleAt),status:row.status as TaskReminder['status'],readAt:nullableStamp(row.readAt),snoozedUntil:nullableStamp(row.snoozedUntil)};
    if(result.status==='read'&&!result.readAt||result.status==='snoozed'&&(!result.snoozedUntil||Date.parse(result.snoozedUntil)<=Date.parse(now)))bad();
    return result;
  });
  if(new Set(items.map(row=>row.task.id)).size!==items.length)bad();
  const total=integer(v.total),unreadCount=integer(v.unreadCount),hasMore=flag(v.hasMore);
  if(total<items.length||hasMore!==(page*40+items.length<total))bad();
  return{items,total,unreadCount,page,pageSize:40,hasMore,serverNow:now,timeZone:'Asia/Shanghai',worker:{lastCheckedAt:nullableStamp(worker.lastCheckedAt),stale:flag(worker.stale),...(worker.capacityBlocked===undefined?{}:{capacityBlocked:flag(worker.capacityBlocked)})}};
}
export function snoozeTime(serverNow:string,choice:'hour'|'tomorrow'):string {
  const now=Date.parse(stamp(serverNow));
  if(choice==='hour')return new Date(now+3600000).toISOString();
  const shanghai=new Date(now+8*3600000),next=Date.UTC(shanghai.getUTCFullYear(),shanghai.getUTCMonth(),shanghai.getUTCDate()+1,1);
  return new Date(next).toISOString();
}
export function reminderIntent(row:TaskReminder,action:'read'|'snooze',serverNow:string,choice:'hour'|'tomorrow'='hour'):ReminderIntent {
  if(!globalThis.crypto?.getRandomValues)throw new Error('请通过安全连接打开看板后再操作。');
  const requestId=Array.from(globalThis.crypto.getRandomValues(new Uint8Array(16)),n=>n.toString(16).padStart(2,'0')).join('');
  return Object.freeze({taskId:row.task.id,requestId,occurrence:row.occurrence,revision:row.revision,action,...(action==='snooze'?{snoozedUntil:snoozeTime(serverNow,choice)}:{})});
}
export function reminderBody(intent:ReminderIntent){const {taskId,...body}=intent;return body;}
export function readReminderOperation(raw:unknown,intent:ReminderIntent):ReminderOperation {
  const v=obj(obj(raw).operation);
  if(v.requestId!==intent.requestId||v.taskId!==intent.taskId||v.occurrence!==intent.occurrence||v.action!==intent.action||v.revision!==intent.revision+1)bad();
  const readAt=nullableStamp(v.readAt),snoozedUntil=nullableStamp(v.snoozedUntil);
  if(intent.action==='read'&&(!readAt||snoozedUntil!==null)||intent.action==='snooze'&&(readAt!==null||!snoozedUntil||Date.parse(snoozedUntil)!==Date.parse(intent.snoozedUntil!)))bad();
  return{...intent,revision:integer(v.revision,1),readAt,snoozedUntil,committedAt:stamp(v.committedAt)};
}

// One tab-scoped minimal envelope. No title, task notes, CSRF or raw identity.
const KEY='family.task-reminder-recovery.v1';
const storageError=()=>new Error('无法保留操作恢复记录，请允许此网站存储后重试。');
const storage=()=>{try{if(typeof sessionStorage==='undefined')throw storageError();return sessionStorage;}catch{throw storageError();}};
export async function reminderScope(identity:string):Promise<string>{
  if(!globalThis.crypto?.subtle)throw storageError();
  const digest=await globalThis.crypto.subtle.digest('SHA-256',new TextEncoder().encode(identity));
  return Array.from(new Uint8Array(digest),n=>n.toString(16).padStart(2,'0')).join('');
}
function readRecovery(raw:unknown):ReminderRecovery {
  const v=obj(raw),i=obj(v.intent);
  if(!['unread','all'].includes(String(v.filter))||!['read','snooze'].includes(String(i.action)))bad();
  const intent:ReminderIntent={taskId:text(i.taskId),requestId:hex(i.requestId,32),occurrence:hex(i.occurrence,64),revision:integer(i.revision),action:i.action as ReminderIntent['action']};
  if(intent.action==='snooze')intent.snoozedUntil=stamp(i.snoozedUntil);else if(i.snoozedUntil!==undefined)bad();
  return{intent:Object.freeze(intent),filter:v.filter as ReminderFilter,page:integer(v.page)};
}
export function loadReminderRecovery(scope:string):ReminderRecovery|null {
  const store=storage();let raw:string|null;
  try{raw=store.getItem(KEY);}catch{throw storageError();}if(!raw)return null;
  let v:Record<string,unknown>;try{v=obj(JSON.parse(raw));}catch{throw storageError();}
  // Called only after /me has confirmed the current member.
  if(v.scope!==scope){try{store.removeItem(KEY);}catch{throw storageError();}return null;}
  try{if(v.version!==1)bad();return readRecovery(v.recovery);}catch{throw storageError();}
}
export function saveReminderRecovery(scope:string,recovery:ReminderRecovery):void {
  const value=JSON.stringify({version:1,scope:hex(scope,64),recovery:readRecovery(recovery)});
  try{const store=storage();store.setItem(KEY,value);if(store.getItem(KEY)!==value)throw storageError();}catch{throw storageError();}
}
export function clearReminderRecovery(scope:string):void {
  try{const store=storage(),raw=store.getItem(KEY);if(raw&&obj(JSON.parse(raw)).scope===scope)store.removeItem(KEY);}catch{throw storageError();}
}
export async function forgetReminderIdentity(identity:string):Promise<void>{clearReminderRecovery(await reminderScope(identity));}
